qa/full-cluster-restart/build.gradle

```diff
@@ -71,6 +71,8 @@ for (Version version : bwcVersions.indexCompatible) {
 
     // some tests rely on the translog not being flushed
     setting 'indices.memory.shard_inactive_time', '20m'
+    // TODO: Use zen2 in bwc tests
+    setting 'discovery.type', 'zen'
```

Comment:
I _think_ these tests should work with Zen2 already. Did you try and find them not to? If not, could you try?

Reply:
I tried these manually and they failed. It seems the metadata is not read correctly.

I'm seeing 

```
[2018-12-06T10:58:30,663][INFO ][o.e.c.c.Coordinator      ] [node-0] setting initial configuration to VotingConfiguration{jxnfx7BsSbuwcz0S8v3psg,soe4_OeBTSGNJm66mOyu0g}
[2018-12-06T10:58:30,803][INFO ][o.e.c.s.MasterService    ] [node-0] elected-as-master ([2] nodes joined)[{node-1}{soe4_OeBTSGNJm66mOyu0g}{c7jlFdFLSiumcLFCWqNjzw}{127.0.0.1}{127.0.0.1:45245}{testattr=test} elect leader, {node-0}{jxnfx7BsSbuwcz0S8v3psg}{hhlkfDtYTE2hZeUKtYFYXg}{127.0.0.1}{127.0.0.1:46175}{testattr=test} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_], reason: master node changed {previous [], current [{node-0}{jxnfx7BsSbuwcz0S8v3psg}{hhlkfDtYTE2hZeUKtYFYXg}{127.0.0.1}{127.0.0.1:46175}{testattr=test}]}, added {{node-1}{soe4_OeBTSGNJm66mOyu0g}{c7jlFdFLSiumcLFCWqNjzw}{127.0.0.1}{127.0.0.1:45245}{testattr=test},}
[2018-12-06T10:58:30,866][INFO ][o.e.c.s.ClusterApplierService] [node-0] master node changed {previous [], current [{node-0}{jxnfx7BsSbuwcz0S8v3psg}{hhlkfDtYTE2hZeUKtYFYXg}{127.0.0.1}{127.0.0.1:46175}{testattr=test}]}, added {{node-1}{soe4_OeBTSGNJm66mOyu0g}{c7jlFdFLSiumcLFCWqNjzw}{127.0.0.1}{127.0.0.1:45245}{testattr=test},}, reason: Publication{term=1, version=2}
[2018-12-06T10:58:31,048][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testhistoryuuidisadded/FNWuN8mXSy6vqEFEU4RkHQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,048][INFO ][o.e.g.DanglingIndicesState] [node-0] [[info/v9CnNlBoQrOFSmx6mKMe9Q]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,048][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testsingledoc/NlUIRH7xSSaT1SECBfMmOA]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testrecovery/OsAKBDv0SmybzcNvsxCLgw]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testrollover-000001/31YMrsVzSlSaZf0ODLYoiA]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testrollover-000002/L1hhzmKJT8SW27-ghSvU2Q]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testshrink/uZHLza9TSAu_U64zfzhRsQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[test_empty_shard/NCCsphB_SvC7pftlKcoegw]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testnewreplicaswork/i1xgSMENTvqDMDCcFR7FeQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testsearch/UWA9LkzgTnqpxwhM1aMiNg]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testsnapshotrestore/1RZJMl1-RumJIJ9BKigC4g]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testshrinkafterupgrade/QbjmPie_SLyQWFxUyKlbxg]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testclusterstate/W4RYzvS_TFqNYfNG1VCR_A]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,050][INFO ][o.e.g.DanglingIndicesState] [node-0] [[queries/gFrWR58uQp6oBH9NWZ2clQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,050][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testshrink_shrunk/VdjBr7yOTYKcvYXZzsv1lQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,054][INFO ][o.e.g.GatewayService     ] [node-0] recovered [0] indices into cluster_state
=========================================
```

and failing tests on the fully upgraded cluster step.

Reply:
The problem here is that the bootstrapping (setInitialConfiguration) wipes the original metadata. Need to fix that.

Reply:
I think this fixes it:

```diff
diff --git a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
index 4869ace4946..0f2b44bdd92 100644
--- a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
+++ b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
@@ -652,7 +652,7 @@ public class Coordinator extends AbstractLifecycleComponent implements Discovery
                 .lastCommittedConfiguration(votingConfiguration)
                 .build();

-            MetaData.Builder metaDataBuilder = MetaData.builder();
+            MetaData.Builder metaDataBuilder = MetaData.builder(currentState.metaData());
             // automatically generate a UID for the metadata if we need to
             metaDataBuilder.generateClusterUuidIfNeeded(); // TODO generate UUID in bootstrapping tool?
             metaDataBuilder.coordinationMetaData(coordinationMetaData);
```

Can't think of a good way to test this apart from this full-cluster-restart test - there are some that check that settings (e.g. snapshot repos) are preserved across the upgrade, which seems good to me.

Reply:
This fix worked and is part of this PR now.